### PR TITLE
Implicit constructor callvalue check

### LINKED ISSUE #8790: Constructors and payable
Inheritance:
 - If base constructor is payable it is possible to inherit even with nonpayable or no constructor at all
 - If base constructor is non payable it is possible to inherit with any kind of constructor
Basically, everything is allowed here, all contracts can be deployed successfully.
Even calling payable base constructor from modifier list in derived constructor is fine

Revert:
 - If base constructor is payable and no constructor in derived contract, passing ether while constructing is fine
 - If base constructor is payable and derived constructor is not payable, it will revert

Setting value for constructor while calling new:
 - If constructor is non payable, value cannot be set with new (it doesn't compile)
 - If there is no constructor, no meter if it has payable or non payable base constructor, it fails compilation

I think we need to define this strictly, especially when inheriting from contracts with payable constructors. If not in old codegen, at least in Sol->Yul.

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/8790

### CODE REVIEW

File: test/libsolidity/semanticTests/various/code_access_runtime.sol [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+
+contract C {
+    function test() public returns (uint256) {
+	D d = new D();
+	bytes32 hash;
+	assembly { hash := extcodehash(d) }
+	assert(hash == keccak256(type(D).runtimeCode));
+	return 42;
```
Comment (ekpyron) [inline test/libsolidity/semanticTests/various/code_access_runtime.sol]: I'd say it doesn't hurt to have this test, even though I didn't end up implementing ``.runtimeCode`` in this PR.

**Comment by chriseth** [comment]
Needs rebase.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 2]
```diff
@@ -159,10 +159,13 @@ void ContractCompiler::appendInitAndConstructorCode(ContractDefinition const& _c
 
 	if (FunctionDefinition const* constructor = _contract.constructor())
 		appendConstructor(*constructor);
-	else if (auto c = _contract.nextConstructor(m_context.mostDerivedContract()))
-		appendBaseConstructor(*c);
 	else
+	{
+		// Implicit constructors are always non-payable.
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp:162]: Can we single this out into its own PR?

**Comment by ekpyron** [comment]
We're missing syntax tests for the errors corresponding to this I think. I'll add some. Based on those we should make double-sure that this really won't seriously break anything :-)!

**Comment by ekpyron** [comment]
Now the Sol->Yul part is extracted and there are more and better tests.

**Comment by chriseth** [comment]
This is either a breaking change or an important bugfix, I would say :)

**Comment by ekpyron** [comment]
Yep, true - I'm not entirely sure which of the two :-).

**Comment by ekpyron** [comment]
Argh... always with the changelog conflicts... that's really annoying :-). I'll move it to "Important Bugfixes" for now, then that won't happen again that often at least.

**Comment by ekpyron** [comment]
We could also decide only to do this with ``sol->yul``... and not at all for old codegen, resp. only for 0.7...

How big do we think is the danger of this causing people to send funds to contracts that will be lost? Is adding the check more dangerous or not having it?

**Comment by chriseth** [comment]
Can you sum up the actual bug again, please?

**Comment by ekpyron** [comment]
> Can you sum up the actual bug again, please?

Sorry, I must have missed that comment until now; still getting used to the new github notifications.

The bug is the following:

Implicit constructors are considered non-payable, i.e. syntactically contracts with such a constructor cannot validly be created with value without low-level functions in inline assembly. However, in the presence of any base constructor (payable or non-payable), the creation code for such contracts did not contain the callvalue check, so creating such contracts with low-level functions did not revert when passing value.

In the absence of any base constructor, the callvalue check *is* generated for implicit constructors, though. Also in the presence of any explicit constructor in the most derived contract the check is generated correctly.

EDIT: I did make the changelog entry a bit more precise as well now.

**Comment by ekpyron** [comment]
So the worst thing that could happen is that someone does the following and expects it to not fail:
```
contract A { constructor() payable {} }
contract B is A {}
contract C {
  function f() public returns (A a) {
    bytes memory c = type(B).creationCode;
    uint value = msg.value;
    assembly { a := create(value, add(c, 0x20), mload(c)) }
  }
}
```
Because this used to work, but will revert after this change. However, ``new B{value: msg.value}()`` is an error for this, so it's only an issue for low-level calls.

The opposite case, i.e. why we should change this at all, is basically the same situation for the case that ``B`` really *doesn't* want to be able to receive ether in the constructor. Funds could be lost because of that, since you can still send it ether with inline assembly without the transaction reverting.

**Comment by ekpyron** [comment]
Seems like we'll be going for treating this as a proper "important bug" with "very low" severity.
Also note: it's not only inline assembly, but also creation transaction that are affected by this.

File: Changelog.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 ### 0.6.8 (unreleased)
 
+Important Bugfixes:
+ * Also add the callvalue check for implicit constructors, if there are base constructors.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Add missing callvalue check to the creation code of a contract that does not define a constructor but has a base that does define a constructor.
```

(I think "implicit constructor" is too vague)

**Comment by chriseth** [comment]
needs an entry in the bug list.

**Review by chriseth — APPROVED** [review]

**Pull request merged**